src/xss-thread-pool.hpp

```diff
@@ -0,0 +1,127 @@
+/*******************************************************************
+ * Copyright (C) 2025 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ * Authors: Raghuveer Devulapalli <raghuveer.devulapalli@intel.com>
+ * ****************************************************************/
+
+#ifndef XSS_THREAD_POOL
+#define XSS_THREAD_POOL
+
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <queue>
+#include <functional>
+#include <future>
+#include <vector>
+#include <atomic>
+
+/*
+ * ThreadPool class and doc: Generated by copilot
+ * This thread pool implementation is a simple and efficient way to manage a
+ * pool of threads for executing tasks concurrently. It uses a std::queue to store
+ * tasks and a set of worker threads which wait for tasks to be added to the
+ * queue. When a task is added, one of the worker threads will pick it up and
+ * execute it. The thread pool can be stopped gracefully, and it also provides
+ * a way to wait for all tasks to complete before stopping.
+ * */
+class ThreadPool {
+private:
+    std::vector<std::thread> workers;
+    std::queue<std::function<void()>> tasks;
+    std::mutex queue_mutex;
+    std::condition_variable condition; // Condition variable for task queue
+    std::condition_variable done_condition; // Condition variable for waiting
+    int active_tasks {0};
+    bool stop;
+
+public:
+    ThreadPool(size_t num_threads) : stop(false)
+    {
+        for (size_t i = 0; i < num_threads; ++i) {
+            // Create a worker thread and add it to the pool
+            // Each thread will run a lambda function that waits for tasks
+            workers.emplace_back([this] {
+                while (true) {
+                    // Lock the queue mutex and wait for a task to be available
+                    std::unique_lock<std::mutex> lock(queue_mutex);
+                    // Wait until there is a task or the pool is stopped
+                    condition.wait(lock,
+                                   [this] { return stop || !tasks.empty(); });
+
+                    // Check if we need to terminate the thread
+                    if (stop && tasks.empty()) { return; }
+
+                    // Extract the next task from the queue
+                    auto task = std::move(tasks.front());
+                    tasks.pop();
+                    // Unlock the mutex before executing the task
+                    lock.unlock();
+                    // Execute the task:
+                    task();
+                }
+            });
+        }
+    }
+
+    template <class F>
+    void enqueue(F &&func)
+    {
+        // Add a new task to the queue and notify one of the worker threads
+        std::unique_lock<std::mutex> lock(queue_mutex);
+        tasks.emplace(std::forward<F>(func));
+        condition.notify_one();
```

Comment:
Race condition: the started threads may not have reached the first lock yet:
```c++
                    // Lock the queue mutex and wait for a task to be available
                    std::unique_lock<std::mutex> lock(queue_mutex);
```
if that's the case, then this `notify_one` will notify no one. That means they will all lock that mutex, then proceed to wait (forever) on the `condition`.

You either want a semaphore or you need to check above before waiting if there already is work to be done.

Reply:
Not sure if I understand this one fully. The `ThreadPool` constructor is called before the call to `qsort_threads` which is well before `submit_task` which in turn calls the `enqueue` method. Is it possible to get 'enqueue` before the `ThreadPool` constructor is called?

Reply:
You're missing the fact that it's two threads. You've described what's happening on the original thread, not on the thread you've just started for the pool. It's entirely possible (and even likely) that the original thread continues running after a couple hundred cycles spent in kernel mode, while the newly started thread takes several thousand cycles to start, in another core (which may be busy). That means the original thread may advance all the way to this function and lock the `queue_mutex` before the started thread has had a chance to lock it and call `wait()`. That means we notify no one here.